unified_planning/io/pddl_writer.py

```diff
@@ -146,31 +146,47 @@ def _write_domain(self, out: IO[str]):
 
         if self.needs_requirements:
             out.write(' (:requirements :strips')
-            if self.problem.kind().has_flat_typing(): # type: ignore
+            if problem_kind.has_flat_typing(): # type: ignore
                 out.write(' :typing')
-            if self.problem.kind().has_negative_conditions(): # type: ignore
+            if problem_kind.has_negative_conditions(): # type: ignore
                 out.write(' :negative-preconditions')
-            if self.problem.kind().has_disjunctive_conditions(): # type: ignore
+            if problem_kind.has_disjunctive_conditions(): # type: ignore
                 out.write(' :disjunctive-preconditions')
-            if self.problem.kind().has_equality(): # type: ignore
+            if problem_kind.has_equality(): # type: ignore
                 out.write(' :equality')
-            if (self.problem.kind().has_continuous_numbers() or # type: ignore
-                self.problem.kind().has_discrete_numbers()): # type: ignore
+            if (problem_kind.has_continuous_numbers() or # type: ignore
+                problem_kind.has_discrete_numbers()): # type: ignore
                 out.write(' :numeric-fluents')
-            if self.problem.kind().has_conditional_effects(): # type: ignore
+            if problem_kind.has_conditional_effects(): # type: ignore
                 out.write(' :conditional-effects')
-            if self.problem.kind().has_existential_conditions(): # type: ignore
+            if problem_kind.has_existential_conditions(): # type: ignore
                 out.write(' :existential-preconditions')
-            if self.problem.kind().has_universal_conditions(): # type: ignore
+            if problem_kind.has_universal_conditions(): # type: ignore
                 out.write(' :universal-preconditions')
-            if (self.problem.kind().has_continuous_time() or # type: ignore
-                self.problem.kind().has_discrete_time()): # type: ignore
+            if (problem_kind.has_continuous_time() or # type: ignore
+                problem_kind.has_discrete_time()): # type: ignore
                 out.write(' :durative-actions')
-            if self.problem.kind().has_duration_inequalities(): # type: ignore
+            if problem_kind.has_duration_inequalities(): # type: ignore
                 out.write(' :duration-inequalities')
             out.write(')\n')
 
-        out.write(f' (:types {" ".join([t.name() for t in self.problem.user_types()])})\n' if len(self.problem.user_types()) > 0 else '') # type: ignore
+        if problem_kind.has_hierarchical_typing(): # type: ignore
+            self.object_freshname = 'object'
```

Comment:
Why storing the object fresh name in self? Creating fields in methods other than `__init__` is quite ugly